user_test.py

```diff
@@ -26,17 +29,83 @@ def tearDown(self):
+            data=data
+        )
+        # Same password and confirm password
+        self.assertIs(data['password'], data['confirm'])
+        with db.connect(DEV_DB_NAME) as db_instance:
+            result = UserModel.select_by_email(db_instance, data['email'])
+            self.assertEqual(result.count, 1)
+            user = result.documents
+            self.assertTrue(self.bcrypt.check_password_hash(user.password, data['password'].encode()))
+
+    def test_register_NotEqualPasswords(self):
+        # Register user
+        data2 = {'email': 'test@test.com', 'password': 'testtest1', 'confirm': 'testfail'}
+        self.test_app.post(
+            '/register/',
+            data=data2
+        )
+        self.assertIsNot(data2['password'],  data2['confirm'])
+
     def test_login(self):
         # Register user
+        data = {'email': 'test@test.com', 'password': 'testtest1', 'confirm': 'testtest1'}
         self.test_app.post(
             '/register/',
-            data={'email': 'test@test.com', 'password': 'testtest1', 'confirm': 'testtest1'}
+            data=data
         )
 
+        # User registered not logged in
+        with db.connect(DEV_DB_NAME) as db_instance:
+            user_collection = db_instance.get_collection(USER_COLLECTION_NAME)
+            result_data = user_collection.find_one({"email": data['email']})
+            self.assertEqual(result_data['_is_authenticated'], False)
+
         # Login user
+        post_data = {'email': 'test@test.com', 'password': 'testtest1'}
         resp = self.test_app.post(
             '/login/',
-            data={'email': 'test@test.com', 'password': 'testtest1'}
+            data=post_data
         )
-        self.assertEqual(resp.status_code, http.HTTPStatus.FOUND)
+        """ tet Db to be used
+        insert first and log in nd test"""
+        # Existing user login with authentication
+        with db.connect(DEV_DB_NAME) as db_instance:
+            result = UserModel.select_by_email(db_instance, post_data['email'])
+            self.assertEqual(result.count, 1)
+            user = result.documents
+            self.assertTrue(self.bcrypt.check_password_hash(user.password, post_data['password'].encode()))
+            user_collection = db_instance.get_collection(USER_COLLECTION_NAME)
+            result_data = user_collection.find_one({"email": post_data['email']})
+            self.assertEqual(result_data['_is_authenticated'], True)
+            self.assertEqual(resp.status_code, http.HTTPStatus.FOUND)
 
+        # Non-existing user login
+        post_data2 = {'email': 'fail@test.com', 'password': 'testtest1'}
+        resp = self.test_app.post(
+            '/login/',
+            data=post_data2
+        )
+        with db.connect(DEV_DB_NAME) as db_instance:
+            result = UserModel.select_by_email(db_instance, post_data2['email'])
+            self.assertEqual(result.count, 0)
+            self.assertNotEqual(resp.status_code, http.HTTPStatus.FOUND)
```

Comment:
Response code 'found' does not seem right to me. Looking at the function, is user does not exist the line that gets executed is:

return render_template('login.html', form=form)

Is that a 302?

It seems it returns 302 when the user exists as in:

return redirect(url_for('main_bp.home'))